# Resolução de um sistema linear a 3 incógnitas - Parte 1

Este notebook faz parte do projeto [Ciência de dados e ML: fundamentos e aplicações.](https://www.kaggle.com/code/regisvargas/ci-ncia-de-dados-e-ml-fundamentos-e-aplica-es)

Neste notebook, iremos abordar o seguinte sistema de equações a 3 incógnitas.

$$
\begin{cases}
2x + 3y - z &= 5 \\
4x - y + 2z &= 6 \\
-x + 5y + 3z &= -4
\end{cases}
$$

A matriz de coeficientes para este sistema é dada por:

In [1]:
import numpy as np

coeficientes = np.array([
    [2,3,-1],
    [4,-1,2],
    [-1,5,3]
], dtype = np.float64)

print(coeficientes)

[[ 2.  3. -1.]
 [ 4. -1.  2.]
 [-1.  5.  3.]]


Vamos agora, calcular o determinante desta matriz.

In [2]:
print(np.linalg.det(coeficientes))

-87.0


O determinante vale $-87$, isto é, o determinante é diferente de zero. [Como já vimos,](https://www.kaggle.com/code/regisvargas/ci-ncia-de-dados-e-ml-fundamentos-e-aplica-es/edit/run/218992775) quando o determinante dos coeficientes de um sistema linear é diferente de zero, o sistema tem uma única solução.

Vamos resolver, passo a passo, este sistema. Vamos rever a matriz de coeficientes.

In [3]:
print(coeficientes)

[[ 2.  3. -1.]
 [ 4. -1.  2.]
 [-1.  5.  3.]]


Podemos também, gerar a matriz de resultados. Veja abaixo.

In [4]:
resultados = np.array([
    [5],
    [6],
    [-4]
], dtype = np.float64)
print(resultados)

[[ 5.]
 [ 6.]
 [-4.]]


Podemos, agora, através do empilhamento horizontal. Criar a matriz que representa o sistema.

In [5]:
sistema = np.hstack((coeficientes, resultados))
print(sistema)

[[ 2.  3. -1.  5.]
 [ 4. -1.  2.  6.]
 [-1.  5.  3. -4.]]


Como sabemos, o primeiro elemento de um array Numpy tem indice $0$. Então, se quisermos retornar a segunda linha da matriz `sistema`, devemos utilizar `sistema[1]`. Veja abaixo.

In [6]:
print(sistema[1])

[ 4. -1.  2.  6.]


Podemos, agora, dividir a segunda linha da matriz por 2.

In [7]:
print(sistema[1]/2)

[ 2.  -0.5  1.   3. ]


E substrair, deste resultado, a primeira linha da matriz `sistema`.

In [8]:
print((sistema[1]/2)-sistema[0])

[ 0.  -3.5  2.  -2. ]


Agora, vamos substituir a segunda linha da matriz `sistema` pela expressão acima.

In [9]:
sistema[1] = (sistema[1]/2)-sistema[0]

A nova matriz `sistema` fica:

In [10]:
print(sistema)

[[ 2.   3.  -1.   5. ]
 [ 0.  -3.5  2.  -2. ]
 [-1.   5.   3.  -4. ]]


Multiplicando a terceira linha da matriz `sistema` por $-2$ ficamos com:

In [11]:
print(-2*sistema[2])

[  2. -10.  -6.   8.]


Subtraindo, deste resultado, a primeira linha temos:

In [12]:
print(-2*sistema[2]-sistema[0])

[  0. -13.  -5.   3.]


Substituindo a terceira linha por esta nova linha, temos:

In [13]:
sistema[2] = -2*sistema[2]-sistema[0]

E a nova matriz `sistema` fica dada por:

In [14]:
print(sistema)

[[  2.    3.   -1.    5. ]
 [  0.   -3.5   2.   -2. ]
 [  0.  -13.   -5.    3. ]]


Multiplicando a terceira linha da matriz `sistema` por $\frac{-3.5}{-13}$, temos:

In [15]:
print((-3.5/-13)*sistema[2])

[ 0.         -3.5        -1.34615385  0.80769231]


Subtraindo, deste resultado, a segunda linha, temos:

In [16]:
print((-3.5/-13)*sistema[2]-sistema[1])

[ 0.          0.         -3.34615385  2.80769231]


Fazendo a substituição, na matriz `sistema", da terceira linha pela linha acima, temos:

In [17]:
sistema[2] = (-3.5/-13)*sistema[2]-sistema[1]

print(sistema)

[[ 2.          3.         -1.          5.        ]
 [ 0.         -3.5         2.         -2.        ]
 [ 0.          0.         -3.34615385  2.80769231]]


Assim, o sistema inicial foi reescrito como:

$$
\begin{cases}
2x + 3y - z &= 5 \\
0x - 3.5y + 2z &= -2 \\
0x + 0y - 3.34615385z &= 2.80769231
\end{cases}
$$

Assim, o valor de $z$ é:

In [18]:
z = 2.80769231/(-3.34615385)
print(z)

-0.83908045949531


Tendo o valor de $z$, calculamos o valor de $y$, por:

In [19]:
y = (-2-2*z)/(-3.5)
print(y)

0.09195402314553716


Tendo o valor de $z$ e de $y$, calculamos $x$ por:

In [20]:
x = (5+z-3*y)/2
print(x)

1.942528735534039


Então, a solução para o sistema apresentado no início do notebook é $(1.942528735534039; 0.09195402314553716; -0.83908045949531)$. O que converge com o resultado encontrado pela função `np.linalg.solve`. Veja abaixo.

In [21]:
print(np.linalg.solve(coeficientes, resultados))

[[ 1.94252874]
 [ 0.09195402]
 [-0.83908046]]


Referências

[Mathematics for Machine Learning and Data Science Specialization](https://www.coursera.org/specializations/mathematics-for-machine-learning-and-data-science)
